### ISSUE TITLE
Simple Injector v3.1.2.0 crashing w3wp.exe in our production environment 

### ISSUE THREAD

**Issue by favill42** [simpleinjector/SimpleInjector#273]
Taken from a .dmp file and analysed using windbg -

> 000000d58b78b858 00007ff9f8e984dd [HelperMethodFrame_1OBJ: 000000d58b78b858] System.Runtime.CompilerServices.RuntimeHelpers.PrepareDelegate(System.Delegate)
000000d58b78b950 00007ff999f08839 SimpleInjector.Internals.CompilationHelpers.JitCompileDelegate(System.Delegate)
000000d58b78b980 00007ff99a515b03 SimpleInjector.Internals.CompilationHelpers.CompileInDynamicAssemblyAsClosure(System.Type, System.Linq.Expressions.Expression, System.Linq.Expressions.ConstantExpression[])
000000d58b78b9e0 00007ff999f0a27d SimpleInjector.Internals.CompilationHelpers.TryCompileInDynamicAssembly(System.Type, System.Linq.Expressions.Expression, System.Delegate ByRef)
000000d58b78ba30 00007ff999f0bdae SimpleInjector.Internals.CompilationHelpers.CompileExpression(System.Type, SimpleInjector.Container, System.Linq.Expressions.Expression, System.Collections.Generic.Dictionary`2)
000000d58b78ba80 00007ff999f0be86 SimpleInjector.Internals.CompilationHelpers.CompileExpression[[System.__Canon, mscorlib]](SimpleInjector.Container, System.Linq.Expressions.Expression)
000000d58b78bad0 00007ff99a515855 SimpleInjector.Registration.BuildDelegate[[System.__Canon, mscorlib]](System.Linq.Expressions.Expression)
000000d58b78bb30 00007ff99a514e08 SimpleInjector.Lifestyles.ScopedRegistration`2[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].BuildExpression()
000000d58b78bb70 00007ff999f0a1aa SimpleInjector.Registration.BuildExpression(SimpleInjector.InstanceProducer)
000000d58b78bbb0 00007ff999f09f80 SimpleInjector.InstanceProducer.BuildExpressionInternal()
000000d58b78bbf0 00007ff9f78e96e4 System.Lazy`1[[System.__Canon, mscorlib]].CreateValue()
000000d58b78bc50 00007ff9f78e92bf System.Lazy`1[[System.__Canon, mscorlib]].LazyInitValue()
000000d58b78bcc0 00007ff999f367ff SimpleInjector.InstanceProducer.BuildExpression()
000000d58b78bd10 00007ff99a5148bf SimpleInjector.ContainerOptions.BuildParameterExpression(System.Type, System.Type, System.Reflection.ParameterInfo)
000000d58b78bd60 00007ff999f0b676 SimpleInjector.Registration.BuildConstructorParameters(System.Type, System.Type, System.Reflection.ConstructorInfo)
000000d58b78be30 00007ff999f0abdc SimpleInjector.Registration.BuildNewExpression(System.Type, System.Type)
000000d58b78be80 00007ff999f0a86d SimpleInjector.Registration.BuildTransientExpression[[System.__Canon, mscorlib],[System.__Canon, mscorlib]]()
000000d58b78bee0 00007ff999f0a1aa SimpleInjector.Registration.BuildExpression(SimpleInjector.InstanceProducer)
000000d58b78bf20 00007ff999f09f80 SimpleInjector.InstanceProducer.BuildExpressionInternal()
000000d58b78bf60 00007ff9f78e96e4 System.Lazy`1[[System.__Canon, mscorlib]].CreateValue()
000000d58b78bfc0 00007ff9f78e92bf System.Lazy`1[[System.__Canon, mscorlib]].LazyInitValue()
000000d58b78c030 00007ff999f367ff SimpleInjector.InstanceProducer.BuildExpression()
000000d58b78c080 00007ff99a5148bf SimpleInjector.ContainerOptions.BuildParameterExpression(System.Type, System.Type, System.Reflection.ParameterInfo)
000000d58b78c0d0 00007ff999f0b676 SimpleInjector.Registration.BuildConstructorParameters(System.Type, System.Type, System.Reflection.ConstructorInfo)
000000d58b78c1a0 00007ff999f0abdc SimpleInjector.Registration.BuildNewExpression(System.Type, System.Type)
000000d58b78c1f0 00007ff999f0a86d SimpleInjector.Registration.BuildTransientExpression[[System.__Canon, mscorlib],[System.__Canon, mscorlib]]()
000000d58b78c250 00007ff999f0a1aa SimpleInjector.Registration.BuildExpression(SimpleInjector.InstanceProducer)
000000d58b78c290 00007ff999f09f80 SimpleInjector.InstanceProducer.BuildExpressionInternal()
000000d58b78c2d0 00007ff9f78e96e4 System.Lazy`1[[System.__Canon, mscorlib]].CreateValue()
000000d58b78c330 00007ff9f78e92bf System.Lazy`1[[System.__Canon, mscorlib]].LazyInitValue()
000000d58b78c3a0 00007ff999f09cc4 SimpleInjector.InstanceProducer.BuildInstanceCreator()
000000d58b78c3f0 00007ff999f09600 SimpleInjector.InstanceProducer.BuildAndReplaceInstanceCreatorAndCreateFirstInstance()
000000d58b78c420 00007ff999f093ff SimpleInjector.InstanceProducer.GetInstance()
000000d58b78c470 00007ff999f08274 SimpleInjector.Container.GetInstanceForRootType(System.Type)
000000d58b78c4b0 00007ff999f081bb SimpleInjector.Container.GetInstance(System.Type)
000000d58b78c4f0 00007ff9e7e8b368 System.Web.Mvc.DefaultControllerFactory+DefaultControllerActivator.Create(System.Web.Routing.RequestContext, System.Type)
000000d58b78c550 00007ff9e7e8aaa4 System.Web.Mvc.DefaultControllerFactory.CreateController(System.Web.Routing.RequestContext, System.String)
000000d58b78c590 00007ff9e7e70714 System.Web.Mvc.MvcHandler.ProcessRequestInit(System.Web.HttpContextBase, System.Web.Mvc.IController ByRef, System.Web.Mvc.IControllerFactory ByRef)
000000d58b78c5f0 00007ff9e7e702f6 System.Web.Mvc.MvcHandler.BeginProcessRequest(System.Web.HttpContextBase, System.AsyncCallback, System.Object)
000000d58b78c650 00007ff9e7e70d5b System.Web.Mvc.HttpHandlerUtil+ServerExecuteHttpHandlerWrapper.Wrap[[System.__Canon, mscorlib]](System.Func`1)
000000d58b78c6a0 00007ff9de3899c9 System.Web.HttpServerUtility.ExecuteInternal(System.Web.IHttpHandler, System.IO.TextWriter, Boolean, Boolean, System.Web.VirtualPath, System.Web.VirtualPath, System.String, System.Exception, System.String)
000000d58b78c780 00007ff9de3894d3 System.Web.HttpServerUtility.Execute(System.Web.IHttpHandler, System.IO.TextWriter, Boolean, Boolean)
000000d58b78c810 00007ff9de389444 System.Web.HttpServerUtility.Execute(System.Web.IHttpHandler, System.IO.TextWriter, Boolean)
000000d58b78c850 00007ff9e7e6fbd0 System.Web.Mvc.Html.ChildActionExtensions.ActionHelper(System.Web.Mvc.HtmlHelper, System.String, System.String, System.Web.Routing.RouteValueDictionary, System.IO.TextWriter)
000000d58b78c8c0 00007ff99a4e88d0 ASP.views_shared_emax_master.__Render__control1(System.Web.UI.HtmlTextWriter, System.Web.UI.Control)
000000d58b78cd40 00007ff9ddc9f437 System.Web.UI.Control.RenderChildrenInternal(System.Web.UI.HtmlTextWriter, System.Collections.ICollection)
000000d58b78cda0 00007ff9ddc9f630 System.Web.UI.Control.RenderControlInternal(System.Web.UI.HtmlTextWriter, System.Web.UI.Adapters.ControlAdapter)
000000d58b78cde0 00007ff9ddc9f3d0 System.Web.UI.Control.RenderChildrenInternal(System.Web.UI.HtmlTextWriter, System.Collections.ICollection)
000000d58b78ce40 00007ff9e7e85dfc System.Web.Mvc.ViewPage.Render(System.Web.UI.HtmlTextWriter)
000000d58b78ce80 00007ff9ddc9f630 System.Web.UI.Control.RenderControlInternal(System.Web.UI.HtmlTextWriter, System.Web.UI.Adapters.ControlAdapter)
000000d58b78cec0 00007ff9ddcaf4d9 System.Web.UI.Page.ProcessRequestMain(Boolean, Boolean)
000000d58b78d200 00007ff9ddcacfe3 System.Web.UI.Page.ProcessRequest(Boolean, Boolean)
000000d58b78d250 00007ff9ddcacefc System.Web.UI.Page.ProcessRequest()
000000d58b78d2a0 00007ff9ddcace76 System.Web.UI.Page.ProcessRequest(System.Web.HttpContext)
000000d58b78d2e0 00007ff9e7e70e30 System.Web.Mvc.HttpHandlerUtil+ServerExecuteHttpHandlerWrapper+c__DisplayClass4.b__3()
000000d58b78d310 00007ff9e7e70d5b System.Web.Mvc.HttpHandlerUtil+ServerExecuteHttpHandlerWrapper.Wrap[[System.__Canon, mscorlib]](System.Func`1)
000000d58b78d360 00007ff9de389ccb System.Web.HttpServerUtility.ExecuteInternal(System.Web.IHttpHandler, System.IO.TextWriter, Boolean, Boolean, System.Web.VirtualPath, System.Web.VirtualPath, System.String, System.Exception, System.String)
000000d58b78d440 00007ff9de3894d3 System.Web.HttpServerUtility.Execute(System.Web.IHttpHandler, System.IO.TextWriter, Boolean, Boolean)
000000d58b78d4d0 00007ff9de389444 System.Web.HttpServerUtility.Execute(System.Web.IHttpHandler, System.IO.TextWriter, Boolean)
000000d58b78d510 00007ff9e7e85f75 System.Web.Mvc.ViewPage.RenderView(System.Web.Mvc.ViewContext)
000000d58b78d580 00007ff9e7e89cfb System.Web.Mvc.ViewResultBase.ExecuteResult(System.Web.Mvc.ControllerContext)
000000d58b78d5f0 00007ff9e7e62f4f System.Web.Mvc.ControllerActionInvoker+c__DisplayClass1a.b__17()
000000d58b78d630 00007ff9e7e62110 System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilter(System.Web.Mvc.IResultFilter, System.Web.Mvc.ResultExecutingContext, System.Func`1)
000000d58b78d690 00007ff9e7e62110 System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilter(System.Web.Mvc.IResultFilter, System.Web.Mvc.ResultExecutingContext, System.Func`1)
000000d58b78d6f0 00007ff9e7e66135 System.Web.Mvc.Async.AsyncControllerActionInvoker+c__DisplayClass25.b__22(System.IAsyncResult)
000000d58b78d740 00007ff9e7e65885 System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeAction(System.IAsyncResult)
000000d58b78d770 00007ff9e7e6b8b8 System.Web.Mvc.Controller+c__DisplayClass1d.b__18(System.IAsyncResult)
000000d58b78d7a0 00007ff9e7e677c6 System.Web.Mvc.Async.AsyncResultWrapper+c__DisplayClass4.b__3(System.IAsyncResult)
000000d58b78d7d0 00007ff9e7e6af14 System.Web.Mvc.Controller.EndExecuteCore(System.IAsyncResult)
000000d58b78d820 00007ff9e7e677c6 System.Web.Mvc.Async.AsyncResultWrapper+c__DisplayClass4.b__3(System.IAsyncResult)
000000d58b78d850 00007ff9e7e6aec5 System.Web.Mvc.Controller.EndExecute(System.IAsyncResult)
000000d58b78d890 00007ff9e7e70b27 System.Web.Mvc.MvcHandler+c__DisplayClass8.b__3(System.IAsyncResult)
000000d58b78d8d0 00007ff9e7e677c6 System.Web.Mvc.Async.AsyncResultWrapper+c__DisplayClass4.b__3(System.IAsyncResult)
000000d58b78d900 00007ff9e7e704a5 System.Web.Mvc.MvcHandler.EndProcessRequest(System.IAsyncResult)
000000d58b78d940 00007ff9ddcb17cf System.Web.HttpApplication+CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
000000d58b78d9c0 00007ff9ddc75c29 System.Web.HttpApplication.ExecuteStep(IExecutionStep, Boolean ByRef)
000000d58b78da10 00007ff9ddc8925b System.Web.HttpApplication+PipelineStepManager.ResumeSteps(System.Exception)
000000d58b78db70 00007ff9ddc76749 System.Web.HttpApplication.BeginProcessRequestNotification(System.Web.HttpContext, System.AsyncCallback)
000000d58b78dbc0 00007ff9ddc6fdfe System.Web.HttpRuntime.ProcessRequestNotificationPrivate(System.Web.Hosting.IIS7WorkerRequest, System.Web.HttpContext)
000000d58b78dc40 00007ff9ddc77c19 System.Web.Hosting.PipelineRuntime.ProcessRequestNotificationHelper(IntPtr, IntPtr, IntPtr, Int32)
000000d58b78ddf0 00007ff9ddc77744 System.Web.Hosting.PipelineRuntime.ProcessRequestNotification(IntPtr, IntPtr, IntPtr, Int32)
000000d58b78de30 00007ff9de35cf4b DomainNeutralILStubClass.IL_STUB_ReversePInvoke(Int64, Int64, Int64, Int32)
000000d58b78e630 00007ff9f8a27fde [InlinedCallFrame: 000000d58b78e630] System.Web.Hosting.UnsafeIISMethods.MgdIndicateCompletion(IntPtr, System.Web.RequestNotificationStatus ByRef)
000000d58b78e630 00007ff9ddce0dde [InlinedCallFrame: 000000d58b78e630] System.Web.Hosting.UnsafeIISMethods.MgdIndicateCompletion(IntPtr, System.Web.RequestNotificationStatus ByRef)
000000d58b78e600 00007ff9ddce0dde DomainNeutralILStubClass.IL_STUB_PInvoke(IntPtr, System.Web.RequestNotificationStatus ByRef)
000000d58b78e6c0 00007ff9ddc77d4b System.Web.Hosting.PipelineRuntime.ProcessRequestNotificationHelper(IntPtr, IntPtr, IntPtr, Int32)
000000d58b78e870 00007ff9ddc77744 System.Web.Hosting.PipelineRuntime.ProcessRequestNotification(IntPtr, IntPtr, IntPtr, Int32)
000000d58b78e8b0 00007ff9de35cf4b DomainNeutralILStubClass.IL_STUB_ReversePInvoke(Int64, Int64, Int64, Int32)
000000d58b78ea88 00007ff9f8a28233 [ContextTransitionFrame: 000000d58b78ea88] 

For your reference our dependencies are defined as follows -

```
this.container = new Container();
this.container.Options.DefaultScopedLifestyle = new WebRequestLifestyle();

this.container.Register<ICachingService, DefaultCachingService>();
this.container.Register<IAuthorizationManagement, AuthorizationManager>();

this.container.RegisterPerWebRequest<IDapperReadRepository>(() => new DapperRepository("ConnectionStringName"));
this.container.RegisterPerWebRequest<IDapperRepository>(() => new DapperRepository("ConnectionStringName"));

this.container.RegisterPerWebRequest<IAwsStorageRepository>(() => AwsStorageRepositoryFactory.Create("Test", "Test"));

container.RegisterMvcControllers(Assembly.GetExecutingAssembly());

DependencyResolver.SetResolver(new SimpleInjectorDependencyResolver(this.container));
```

This issue is effecting our production servers - its causing the w3wp process to terminate and restart - would you be able to assist me please?

**Comment by dotnetjunkie** [comment]
**TLDR;**

Set this configuration switch:

``` c#
container.Options.EnableDynamicAssemblyCompilation = false;
```

More information can be found [here](https://stackoverflow.com/questions/38134310/container-verify-throws-system-executionengineexception-sometimes).

**Comment by favill42** [comment]
Hi dotnetjunkie,

Thanks for the super quick response :+1: 

I've checked our production servers and we have no monitoring setup

**Comment by dotnetjunkie** [comment]
These issues are usually caused by limitations (or bugs) in tools that we use. Those tools are sometimes unable to cope with dynamic assembly compilation. Finding out what tool causes it and how to prevent this can often be a lot of work, and if it's caused by a bug or limitation of the tool, it can be quite daunting to communicate with the provider of the tool (sometimes Microsoft itself) to solve this.

So as a quick fix, flip the `EnableDynamicAssemblyCompilation` switch to `false`, since this will almost certainly fix the problem for you. The performance penalty you get by this is very low, and is not something you'll have to be concerned about. We are even considering to change the default to `false` in a future release anyway and possibly even remove the feature completely in a next major release.

Flip the switch!

**Comment by favill42** [comment]
The switch has been flipped. We'll see what tomorrow brings...

**Comment by dotnetjunkie** [comment]
I consider the issue fixed.

**Comment by favill42** [comment]
Yes its fixed now, thank you
